### PR TITLE
Fix #417 by adding an add user route, local action and form

### LINKED ISSUE #417: Add user to group
I don't see any way for a site admin to create a group and add several users as admins and members of that group. On behalf of those users. Not by asking them to personally subscribe to the group.

### PR DESCRIPTION
**Pull request by heddn**

### CODE REVIEW

File: src/Form/GroupUserSubscribeForm.php [hunk 1]
```diff
@@ -0,0 +1,127 @@
+<?php
+
+namespace Drupal\og\Form;
```
Comment (MPParsley) [inline src/Form/GroupUserSubscribeForm.php:3]:
```suggestion
declare(strict_types = 1);

namespace Drupal\og\Form;
```

File: src/Plugin/Derivative/OgLocalAction.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
+<?php
+
+namespace Drupal\og\Plugin\Derivative;
```
Comment (MPParsley) [inline src/Plugin/Derivative/OgLocalAction.php:3]:
```suggestion
declare(strict_types = 1);

namespace Drupal\og\Plugin\Derivative;
```

File: src/Form/GroupUserSubscribeForm.php [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    foreach (OgRole::loadByGroupType($membership->getGroup()->getEntityTypeId(), $membership->getGroup()->bundle()) as $role) {
+      // Only add the role to the list if it is not a required role, these
+      // cannot be added. Nor should invalid roles be added.
+      if (!$role->isRequired() && $membership->isRoleValid($role)) {
+        $options[$role->id()] = $role->label();
+      }
+    }
+    $form['roles'] = [
+      '#type' => 'select',
+      '#title' => t('Roles'),
```
Comment (MPParsley) [inline src/Form/GroupUserSubscribeForm.php]:
```suggestion
      '#title' => $this->t('Roles'),
```

File: src/Form/GroupUserSubscribeForm.php [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    parent::submitForm($form, $form_state);
+    /** @var \Drupal\og\OgMembershipInterface $membership */
+    $membership = $this->entity;
```
Comment (bboro) [inline src/Form/GroupUserSubscribeForm.php:121]: Thanks for the useful PR! I tried it and the roles aren't getting saved. So we need to save the roles here:

```suggestion
    $membership = $this->entity;
    $assigned_roles = $form_state->getValue('roles');
    if (!empty($assigned_roles)) {
      $roles = OgRole::loadMultiple($assigned_roles);
      foreach($roles as $role) {
        $membership->addRole($role);
      }
    }
```

Comment (MPParsley) [reply src/Form/GroupUserSubscribeForm.php:121]: Not sure what went wrong here but roles is a basefield on the membership entity so it should get saved.
```
    $fields['roles'] = BaseFieldDefinition::create('entity_reference')
      ->setLabel(t('Roles'))
      ->setDescription(t('The OG roles related to an OG membership entity.'))
      ->setCardinality(FieldStorageDefinitionInterface::CARDINALITY_UNLIMITED)
      ->setSetting('target_type', 'og_role')
      ->setSetting('handler', 'og:og_role')
      ->setConstraints(['ValidOgRole' => []])
      ->setDisplayOptions('form', [
        'type' => 'options_buttons',
        'weight' => 0,
      ])
      ->setDisplayConfigurable('view', TRUE)
      ->setDisplayConfigurable('form', TRUE);
```

Are you on the dev branch?

Comment (bboro) [reply src/Form/GroupUserSubscribeForm.php:121]: I was wondering about that, I'm on `1.0.0-alpha5`. If it's already resolved false alarm then! Thanks

**Comment by amitaibu** [comment]
@bboro care to add tests to this, so we could get it in? :smile_cat:

**Comment by bboro** [comment]
@heddn Hi!

I wrote some tests for this PR and they are currently in https://github.com/heddn/og/pull/2 could you review it when you have a chance. It looks like we'll need to reroll the branch as well against current 1.x, so let me know if you'd like me to take over this instead, and I can reroll + open a new PR from my fork.

Cheers

**Comment by MPParsley** [comment]
Closing in favor of #231.

**Pull request closed without merging**